MRG: Faster reading

Eric89GXL opened the pull request:
This should universally speed up our FIF reading functions at least a little bit. On `master` `test_read_write_epochs` takes ~15 sec and on this PR it takes ~13. So a small improvement, but worthwhile. It also makes the code a little bit easier to parse.

Eric89GXL commented:
(I was prompted to do this because I saw Travis had really long run times for tests that are reading-heavy. This might help a little bit.)

dengemann commented on mne/io/tag.py:
```diff
@@ -203,6 +204,257 @@ def _loc_to_eeg_loc(loc):
+    d['kind'] = int(np.fromstring(fid.read(4), dtype=">i4"))
+    d['range'] = float(np.fromstring(fid.read(4), dtype=">f4"))
+    d['cal'] = float(np.fromstring(fid.read(4), dtype=">f4"))
+    d['coil_type'] = int(np.fromstring(fid.read(4), dtype=">i4"))
```
we have a lot of those ...

why not
```Python
d = dict(
    scanno=int(np.fromstring(fid.read(4), dtype=">i4")),
    logno=int(np.fromstring(fid.read(4), dtype=">i4"))
)  # and so forth
```

jona-sassenhagen replied:
Or loop/dict comp here, like

```
d = dict(key, int(np.fromstring(fid.read(4), dtype=">i4")) 
         for key in ('scanno', 'logno', ...))  # or however dict comp works in py2.7, I keep forgetting
for key in ('range', 'cal'):
    d[key] = float(np.fromstring(fid.read(4), dtype=">i4"))
```

Eric89GXL replied:
same with this -- I could use the below pattern, but @dengemann's is slightly more readable I think, even though it's less DRY

jona-sassenhagen replied:
The argument in favour of looping is you know at once that one procedure is used for all inputs (or here, you know there are exactly two processes for two sets of vars). I'm not sure this is more readable though myself either.

jona-sassenhagen replied:
I.e., here, you have to actually parse every line to see that some are int, some are float, otherwise they're the same etc.

Eric89GXL replied:
@jona-sassenhagen I tried implementing this but it ends up looking a bit clunky since sometimes we cast, other times we don't. So I'm going to go with @dengemann's idea.

dengemann commented on mne/io/tag.py:
```diff
@@ -203,6 +204,257 @@ def _loc_to_eeg_loc(loc):
+    return d
+
+
+def _read_old_pack(fid, tag, shape, rlims):
```
no mini doc strings here and there

dengemann replied:
forget about it just read your comment.

Eric89GXL replied:
@jasmainak asked and answered :)

jasmainak commented on mne/io/tag.py:
```diff
@@ -203,6 +204,257 @@ def _loc_to_eeg_loc(loc):
+    #
+    #   Read the coil coordinate system definition
+    #
+    # deal with nasty OSX Anaconda bug by casting to float64
```
what exactly is the bug? Can you put a link here?

Eric89GXL replied:
I don't know the bug actually, this line is more or less copy pasted

jasmainak replied:
I believe you are the original author :)

Eric89GXL replied:
Looks like from back in 2013... have no idea what that was :)

jasmainak commented:
How did you time the test? I have noticed that they differ from run to run depending on what processes run in the background

jasmainak commented:
changing 400 lines of code to decrease test time by 2 s? Shouldn't we try to get the low-hanging fruit first? 

I think the interpolation test can be made faster quite easily. We are testing the same thing (correlation > 0.8) for epochs, raw, and evoked. If we precompute the mapping matrix and just use it for all three, we can cut down test time by a factor of three (I am guessing?)

jasmainak commented:
I'm -1 on merging this. If testing time is the problem, the test should be fixed and refactored rather than changing 400 lines of code.

Maybe this is a crazy idea, but if IO is the bottleneck can't we try to test by writing to this type of file: https://docs.python.org/2/library/tempfile.html?highlight=tempfile#tempfile.SpooledTemporaryFile

Eric89GXL commented:
> I'm -1 on merging this. If testing time is the problem, the test should
be fixed and refactored rather than changing 400 lines of code.

To me it's worth improving read performance by even 1% if we can do so
without making the code more complex (and the gains here can be better than
that). FIFF reading is used all the time, so optimizing it is useful beyond
just test times, even though that's what prompted me to start working on
it. In this instance, I think the changes make the code more readable, not
less. It also makes our reading more similar to how writing works --
instead of a single monolithic function, operations get delegated to
subfunctions. I believe this is a bit better of a design.

I think the tests should also be refactored for speed, but I don't think
that's a good reason to reject speed gains and better readability through
these changes.

I timed things by using line profiling to see that a non trivial amount of
time was spent triaging which writing functions to use, especially when
reading a measurement info block (which we use very often). The dict access
implemented here fixes that.

Eric89GXL commented:
> I timed things by using line profiling to see that a non trivial amount of time was spent triaging which writing functions to use

To provide another example, in `test_read_write_epochs`, on `master` 4.5 seconds are spent in `read_meas_info`. On my PR, it's 3 seconds.

@dengemann @agramfort do you also agree these changes are not worthwhile?

jaeilepp commented:
To me this just looks like refactored code with speed gains on the side (although from the diff it's quite hard to compare). +1 from me.

jasmainak commented:
okay, if Jaakko is motivated enough to properly review the full diff, it is fine with me. I am concerned about slight incremental gain for changing so many lines of code (on the first commit itself). What happens is that people end up not reviewing the code but blindly trusting the author. I am still -1

jasmainak commented:
> In this instance, I think the changes make the code more readable, not less. It also makes our reading more similar to how writing works -- instead of a single monolithic function, operations get delegated to subfunctions. I believe this is a bit better of a design.

So, maybe you are talking about two separate PRs, not one.

jona-sassenhagen commented:
Does it matter that much if this one is called "Faster reading" rather than "Refactor reading (also a bit of speed-up)"?

To me the code seems a good bit more readable with the refactoring - if it also gives performance boosts, that's good.

jasmainak commented:
@jona-sassenhagen please review then :)

Eric89GXL commented:
> So, maybe you are talking about two separate PRs, not one.

I'm not sure what the two separate PRs would be in this case. Once the functions are separated out (the vast majority of the changed lines are a result of this), we need some way to call the proper one. Putting it into a dict is a natural choice instead of leaving a bunch of conditionals.

jona-sassenhagen commented on mne/io/tag.py:
```diff
@@ -48,16 +53,12 @@ def __repr__(self):
+                   self.size == tag.size and
+                   self.next == tag.next and
+                   self.pos == tag.pos and
+                   self.data == tag.data)
```
Could possibly be improved by looping over a list of str and using getattr?

Eric89GXL replied:
Maybe, but I'm not sure that ends up being more readable

jasmainak replied:
yeah, I agree with Eric here

jasmainak commented on mne/io/tag.py:
```diff
@@ -203,6 +204,257 @@ def _loc_to_eeg_loc(loc):
+    return data
+
+
+def _read_simple(fid, tag, shape, rlims, dtype):
```
must have a minimal docstring even if it's a private function.

Eric89GXL replied:
see comment above (https://github.com/mne-tools/mne-python/pull/2865/files#diff-74145b1b9e6710714418f4ed21d552cbR210)

jasmainak replied:
It still is a convention we have been following. To put one line docstrings even if you don't describe the params. Are we abandoning the convention or is this an exception?

Eric89GXL replied:
Just making an exception. The policy is in place presumably to make code readable and usable -- in this case I think the docstrings wouldn't help that much. But it's not too bad to add them if you think they would help.

jasmainak replied:
okay cool

dengemann replied:
Thanks @jasmainak reminding of core patterns-- we need this to stay consistent. Let's allow for some flexibility if someone experienced insists with on having some local patterns with a reason-- we need this to not become bureaucrats.

jasmainak replied:
lol bureaucrats ;) My main objection was that we should be careful of making PRs which are hard to review. Eric posted a PR which said that it made IO faster. There was one line in the diff which claimed to make something faster. Was that it? I'm still in the dark as to what made it faster. Anyway, I trust Eric with not introducing bugs so no issues :)

Eric89GXL replied:
FYI if you look at the final diff I did end up adding the one-line docstrings, so the bureaucracy won in the end :)

> I'm still in the dark as to what made it faster. 

Looking back in the (long) thread a bit there is this nugget:

> I timed things by using line profiling to see that a non trivial amount of time was spent triaging which writing functions to use, especially when reading a measurement info block (which we use very often). The dict access implemented here fixes that.

[I just made a gist to demonstrate the idea] (https://gist.github.com/Eric89GXL/be9052ff4cf602796599). To elaborate, a conditional with 18 options, which is what the function used to use, that gets accessed 500 times (e.g., our `test-ave.fif` has 673 tags, the vast majority of which are part of `Info`) will take ~2.5-3x longer to evaluate its conditionals than an equivalent `dict.get()`-based method. This is because `dict` uses hashing for access which can be very fast, whereas a bunch of `elif` clauses must be interpreted and thus will be slow if it's used a lot. This is what took the `read_meas_info` time in that test from 4.5 to 3.0 seconds for me. The bulk of the time will still be spent actually *reading* the data as opposed to choosing the function to read the data, but this at least cuts down on the latter overhead.

jasmainak replied:
okay, got it. Very cool gist :) No idea how I missed that in the original comment. I think there are other parts of our code which might benefit from this too.

jasmainak commented on mne/io/tag.py:
```diff
@@ -203,6 +204,257 @@ def _loc_to_eeg_loc(loc):
+    return data
+
+
+def _read_coord_trans_struct(fid, tag, shape, rlims):
```
same comment for all the private functions.

jasmainak commented:
@jaeilepp @jona-sassenhagen did you manage to find _which_ lines exactly are making the IO faster :)

jasmainak commented on mne/io/tag.py:
```diff
@@ -230,253 +482,22 @@ def read_tag(fid, pos=None, shape=None, rlims=None):
     s = fid.read(4 * 4)
+    # This is faster than np.fromstring, saves ~10% of time in some places (!)
+    s = struct.unpack('>iIii', s)
+    # s = np.fromstring(s, dtype='>i4,>u4,>i4,>i4')[0]
```
remove

Eric89GXL commented:
Comments addressed by latest commit. Ready for review/merge from my end.

dengemann commented:
@Eric89GXL thanks for doing this! I think any second that we can gain on average is good. It could make  a difference that let's you pay one hour less to your favourite CPU rental service.

agramfort commented:
+1 for merge if travis error is unrelated

thx @Eric89GXL

dengemann commented:
Thanks Eric!

jasmainak commented:
Thanks! :)
